### PR TITLE
Use the MedusaConfiguration API when creating Medusa configuration

### LINKED ISSUE #1158: Add a field to reference a MedusaConfiguration resource in the K8ssandraCluster API
The Medusa spec would contain an additional [ObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/object-reference/) field to set the reference to the MedusaConfiguration object:

```yaml
  medusa:
    medusaConfigurationRef:
      name: medusa-storage1
      namespace: ns1
      api: v1alpha1
      kind: MedusaConfiguration
    storageProperties:
      prefix: my-cluster
      transferMaxBandwidth: 150MB/s
```

The above would generate Medusa storage settings [here](https://github.com/k8ssandra/k8ssandra-operator/blob/main/pkg/medusa/reconcile.go#L56), with the data stored in the medusa-storage-bucket1 object, and override the prefix and transfer max bandwidth with the provided values.
Merging of both structures would be done with Goalesce, prioritizing the values from storageProperties.
The CreateMedusaIni() function would be refactored to take a MedusaSpec object instead of a K8ssandraCluster one, which would be the result of the above merge.

An additional check would make sure the prefix is explicitly set in storageProperties when a medusaConfigurationRef is set. This will reduce the odds of putting backups from multiple clusters in the same folder.

```[tasklist]
### Definition of Done
- [x] The K8ssandraCluster spec has a new optional `medusa.medusaConfigurationRef` field which can point to an existing MedusaConfiguration resource
- [x] During the reconcile, the controller will merge the MedusaConfiguration object fields with the storageProperties ones, using Goalesce and prioritizing the latter
- [x] The prefix is mandatory in the storageProperties when using a MedusaConfiguration ref
```

### PR DESCRIPTION
**Pull request by rzvoncek**
**What this PR does**:
* Adds an ObjectReference to the Medusa spec to allow referencing MedusaConfig.
* Modifies the k8ssandra cluster creation webohook to error out if the reference is present but the medusa storage prefix is not set in the cluster's spec.
* When creating the cluster, or rather it's medusa.ini config map, merge both the MedusaConfig and the cluster spec, giving priority to the later.

**Which issue(s) this PR fixes**:
Fixes #1158.

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CHANGELOG.md updated (not required for documentation PRs)
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Review by adejanovski — CHANGES_REQUESTED** [review]
A few minor changes and a **todo:**

We need an e2e test for this.
Modify the `single-dc-encryption-medusa` fixture to add a MedusaConfiguration object and modify the K8ssandraCluster storage properties to replace them with the medusaConfigurationRef.
The medusa-bucket-key will be created already (it's at the root of the `fixtures` directory) so you just need to reference it in the MedusaConfiguration object.

File: apis/k8ssandra/v1alpha1/k8ssandracluster_webhook.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ var (
 	ErrNoStorageConfig = fmt.Errorf("storageConfig must be defined at cluster level or dc level")
 	ErrNoResourcesSet  = fmt.Errorf("softPodAntiAffinity requires Resources to be set")
 	ErrClusterName     = fmt.Errorf("cluster name can not be changed")
+	ErrNoStoragePrefix = fmt.Errorf("medusa storage prefix must be set")
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply apis/k8ssandra/v1alpha1/k8ssandracluster_webhook.go:38]: Ammended into the initial commit.

File: controllers/k8ssandra/medusa_reconciler.go [hunk 2]
```diff
@@ -32,6 +35,12 @@ func (r *K8ssandraClusterReconciler) reconcileMedusa(
 	if medusaSpec != nil {
 		logger.Info("Medusa is enabled")
 
+		desiredKc := kc.DeepCopy()
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/k8ssandra/medusa_reconciler.go:38]: Ammended into the inital commit.

File: controllers/k8ssandra/medusa_reconciler_test.go [hunk 3]
```diff
@@ -387,3 +436,105 @@ func reconcileMedusaStandaloneDeployment(ctx context.Context, t *testing.T, f *f
...
+	// add a reference to the MedusaConfiguration object to the cluster spec
+	kcSecondAttempt.Spec.Medusa.MedusaConfigurationRef.Name = medusaConfigName
+
+	// confirm the prefix in the storage properties is still empty
+	require.Equal("", kcSecondAttempt.Spec.Medusa.StorageProperties.Prefix)
+
+	// retry creating the cluster
+	t.Log("Creating k8ssandracluster with Medusa and MedusaConfig but without a prefix in the cluster spec")
+	err = f.Client.Create(ctx, kcSecondAttempt)
+	require.Error(err, "creating a cluster without Medusa's storage prefix should not happen if if the MedusaConfig object exists")
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/k8ssandra/medusa_reconciler_test.go:536]: Ammended into an exisitng commit.

File: docs/content/en/tasks/backup-restore/_index.md [hunk 4]
```diff
@@ -131,6 +131,28 @@ This allows creating bucket configurations that are easy to share across multipl
...
+metadata:
+  name: demo
+spec:
+  cassandra:
+    ...
+  medusa:
+    medusaConfigurationRef:
+      name: medusaconfiguration-s3
+    storageProperties:
+      peffix: demo
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply docs/content/en/tasks/backup-restore/_index.md:152]: Ammended into the initial commit.

**Review by adejanovski — APPROVED** [review]

**Pull request merged**